File: interpreters/typescript/test/corpus.test.ts
/**
 * WHY THIS FILE EXISTS: the comparison harness. Every published example gets
 * this implementation's verdict compared against a committed expectation, so
 * "the TypeScript agrees with the Rust" is a table a reviewer can read rather
 * than a claim in a report. `testkit/corpus.ts` holds the table and the WHY for
 * each row.
 *
 * WHAT IT PINS: the stated verdict for every row present on disk, AND — the
 * half that is easy to leave out — that no `*.json` in `examples/` lacks a row.
 * That second direction is deliberately a COVERAGE check rather than a count:
 * the corpus grows, and a number written into prose is a fact that rots
 * silently in every place someone forgot to look. A file nobody classified
 * fails here and names itself.
 */

import assert from 'node:assert/strict';
import { test } from 'node:test';

import { AphError } from '../src/errors.js';
import { isDidKey } from '../src/didkey.js';
import { parseEnvelope } from '../src/parse.js';
import { verifyProofStructure } from '../src/structure.js';
import { verifyEnvelope, verifyProofAt } from '../src/verify.js';
import { proofsOf, type EnvelopeProof } from '../src/types.js';
import type { SuppliedKeys } from '../src/verify.js';
import { CORPUS, exampleExists, listExampleFiles, readExample } from '../testkit/corpus.js';
import {
  GOLDEN_EVALUATION_INSTANT,
  GOLDEN_FILE,
  goldenSuppliedKeys,
} from '../testkit/golden.js';
import { TS_MINTED_EVALUATION_INSTANT, TS_MINTED_FILE } from '../testkit/ts_minted.js';

/**
 * The instant each file is evaluated at. Every published envelope carries its
 * own 24-hour window, so one shared constant would put half the corpus outside
 * it — and a verifier that read the wall clock would pass today and fail next
 * year, which is the whole reason `now` is a parameter.
 */
function evaluationInstant(file: string): string {
  return file === TS_MINTED_FILE ? TS_MINTED_EVALUATION_INSTANT : GOLDEN_EVALUATION_INSTANT;
}

/**
 * Keys supplied per file. Only the Rust golden needs any: its notary is a
 * `did:web`. Everything else in the corpus resolves offline, which is the
 * property that makes a corpus checkable by a stranger.
 */
function suppliedKeys(file: string): SuppliedKeys | undefined {
  return file === GOLDEN_FILE ? goldenSuppliedKeys() : undefined;
}

for (const row of CORPUS) {
  test(`corpus: ${row.file} — ${row.verdict.kind}`, async (t) => {
    if (!exampleExists(row.file)) {
      assert.equal(
        row.required,
        false,
        `${row.file} is a REQUIRED corpus row and is missing from examples/`,
      );
      t.skip(
        `${row.file} is not on disk yet — an optional row for a vector this repository ` +
          `publishes that is not committed yet. It proves: ${row.proves}`,
      );
      return;
    }

    const text = readExample(row.file);
    const options = {
      now: evaluationInstant(row.file),
      keys: suppliedKeys(row.file),
    };

    if (row.verdict.kind === 'admit') {
      const verified = await verifyEnvelope(text, options);
      assert.ok(verified.envelope.id.length > 0);
      return;
    }

    if (row.verdict.kind === 'refuse') {
      const expected = row.verdict.code;
      try {
        await verifyEnvelope(text, options);
      } catch (error) {
        if (!(error instanceof AphError)) {
          assert.fail(`${row.file} must be refused with a §11 code, got ${String(error)}`);
        }
        assert.equal(error.code, expected);
        return;
      }
      assert.fail(`${row.file} was ADMITTED but the table expects ${expected}`);
      return;
    }

    // `offline-proofs-only`: parse and structure must hold, and every proof
    // whose key travels inside its own identifier must verify. The rest is
    // unchecked and the row says so.
    const envelope = parseEnvelope(text);
    verifyProofStructure(envelope);
    const proofs = proofsOf(envelope);
    let checked = 0;
    for (let index = 0; index < proofs.length; index += 1) {
      const proof = proofs[index] as EnvelopeProof;
      if (!isDidKey(proof.verificationMethod)) continue;
      assert.equal(
        await verifyProofAt(envelope, index, {}),
        true,
        `${row.file} proof ${index} names a did:key and must verify offline`,
      );
      checked += 1;
    }
    assert.ok(
      checked > 0,
      `${row.file} has no did:key proof, so this row checks nothing — give it an admit or ` +
        'refuse verdict with the key it needs',
    );
  });
}

test('every example on disk has a row in the corpus table', () => {
  // The direction that catches the thing a count never does: a vector landing
  // in `examples/` with nobody having decided what this implementation should
  // say about it.
  const classified = new Set(CORPUS.map((row) => row.file));
  const unclassified = listExampleFiles().filter((file) => !classified.has(file));
  assert.deepEqual(
    unclassified,
    [],
    `these files are in examples/ with no corpus row: ${unclassified.join(', ')}`,
  );
});

test('every corpus row states WHY its file is in the corpus', () => {
  // A row with no reason is a row nobody can maintain: the next person cannot
  // tell whether a changed verdict is a regression or a correction.
  for (const row of CORPUS) {
    assert.ok(row.proves.length > 40, `${row.file} has no meaningful WHY`);
  }
});

File: interpreters/elixir/test/corpus_test.exs
defmodule APH.CorpusTest do
  use ExUnit.Case, async: true

  test "the corpus on disk is exactly the corpus the manifest claims" do
    # WHY: this suite read exactly two published examples BY NAME and
    # enumerated nothing, so a vocabulary change — a new channel shape, a new
    # cryptosuite vector, a renamed file — could land in `examples/` and this
    # binding would report success having never opened it. A count would not
    # have helped: a floor of "at least twelve" passes forever, and swapping one
    # file for another leaves a count unmoved.
    #
    # PINS: SET EQUALITY in BOTH directions between the manifest's conformance
    # list and the top-level `*.json` files on disk. A file on disk with no
    # manifest entry fails and names itself — the direction that catches a
    # vector nobody classified. A manifest entry with no file fails too, which
    # catches a deletion or rename that a one-directional check would read as
    # "fewer files, still above the floor".
    on_disk = APH.TestCorpus.example_files()
    claimed = APH.TestCorpus.conformance_files()

    undeclared = on_disk -- claimed
    missing = claimed -- on_disk

    assert undeclared == [],
           "these files are in examples/ with no entry in manifest.json: " <>
             Enum.join(undeclared, ", ")

    assert missing == [],
           "manifest.json claims these files that are not on disk: " <> Enum.join(missing, ", ")
  end

  test "every conformance file the manifest claims strict-parses through the NIF" do
    # WHY: the set comparison above compares NAMES. A file can be named in both
    # places and still be a zero-byte stub — and an audit found the old floor
    # here ("is a JSON object") let a golden carrying a NEW FIELD pass without
    # the NIF ever seeing the field, so the one gate that runs this boundary
    # was vouching for bytes it never parsed.
    #
    # PINS: every conformance entry strict-parses through APH.parse_envelope_json/1,
    # so every FUTURE golden exercises the NIF the day it lands, with nobody
    # remembering to add a test.
    for name <- APH.TestCorpus.conformance_files() do
      raw = APH.TestCorpus.read!(name)

      assert {:ok, _normalized} = APH.parse_envelope_json(raw),
             "#{name} is named in manifest.json and does not strict-parse through the NIF"
    end
  end

  test "every excluded file is on disk and says why it is excluded" do
    # WHY: the excluded list is the half of the inventory that rots quietly. A
    # conformance claim that silently covered a deliberately non-conformant
    # document would be false, and an exclusion naming a file nobody can find
    # any more is an exclusion for something that stopped existing.
    #
    # PINS: every excluded path resolves on disk, and every one carries a
    # non-empty reason — an exclusion with no stated reason is one the next
    # reader cannot tell from an oversight.
    for entry <- APH.TestCorpus.excluded_entries() do
      path = Map.fetch!(entry, "path")
      reason = Map.get(entry, "reason", "")

      assert String.trim(reason) != "", "#{path} is excluded with no stated reason"

      assert File.exists?(Path.join(APH.TestCorpus.examples_dir(), path)),
             "#{path} is excluded in manifest.json and is not on disk"
    end
  end

  test "the two deep-verified goldens are still the files the manifest claims" do
    # WHY: the named goldens are the ONE place this suite still addresses the
    # corpus by hand — `principal_signed_golden/0` and `legacy_slack_reply/0`
    # carry every cryptographic and boundary assertion in the other test files.
    # A rename that updated the manifest and the disk together would leave those
    # two helpers pointing at nothing, and the failure would surface as a
    # File.read! deep inside an unrelated test.
    #
    # PINS: both hand-named goldens are members of the declared conformance set,
    # so the individual names and the set inventory cannot drift apart.
    claimed = APH.TestCorpus.conformance_files()

    for name <- ["principal_signed_envelope.json", "slack_reply_envelope.json"] do
      assert name in claimed,
             "#{name} is read by name throughout this suite and is not in the conformance manifest"
    end

    assert String.contains?(APH.TestCorpus.principal_signed_golden(), "\"proof\"")
    assert String.contains?(APH.TestCorpus.legacy_slack_reply(), "\"proof\"")
  end
end
